rfcs/convergence/props-typing-improvements.md

```diff
@@ -0,0 +1,148 @@
+
+```typescript
+export const useWidget = (props: WidgetProps, ref: React.Ref<HTMLElement>, defaultProps?: WidgetProps): WidgetState => {
+  const state = mergeProps(
+    {
+      as: 'div',
+      test: { as: 'span' },
+      propWithDefaultValue: 'hello world',
+    },
+    defaultProps,
+    resolveShorthandProps(props, widgetShorthandProps),
+  );
+
+  return state;
+};
+```
+
+**Problem:** in hooks like `useWidgetStyles`, there's either an implicit assumption about default props, or unnecessary checking is performed:
+
+```typescript
+// Unchecked assumption that the value is not undefined:
+[s => s.propWithDefaultValue!.contains('hello'), {/* ... */}],
+// Unnecessary undefined check:
+[s => s.propWithDefaultValue && s.propWithDefaultValue.contains('hello'), {/* ... */}],
+```
+
+## Detailed Design or Proposal
+
+The fix would be to:
+
+1. Add some helper types
+   - `ComponentState`: assists in creating the `State` type from the `Props` type: the dev simply lists the names of the props that are slot props, and the props that have default values.
+   - `ComponentShorthandProps`: a list of the shorthand props from the state.
+2. Update `mergeProps` (`makeMergeProps`) to require that the first argument is of type `TState`.
+   - This will enforce that all props with defaults are provided and are of the correct type.
+
+### Usage example
+
+`Widget.types.ts`:
+
+```typescript
+export type WidgetState = ComponentState<
+  WidgetProps,
+  { ShorthandProps: 'icon' | 'text' },
+  { DefaultedProps: 'text' | 'propWithDefaultValue' }
+>;
+```
+
+`useWidget.ts`:
+
+The types are checked, and the bug is fixed! (there was a typo in the problem statement above: the slot name was `test` instead of `text`)
+
+```typescript
+export const widgetShorthandProps: ComponentShorthandProps<WidgetState>[] = ['icon', 'text'];
+
+export const useWidget = (props: WidgetProps, ref: React.Ref<HTMLElement>, defaultProps?: WidgetProps): WidgetState => {
+  const state = mergeProps(
+    {
+      as: 'div',
+      text: { as: 'span' },
+      propWithDefaultValue: 'hello world',
+    },
+    defaultProps,
+    resolveShorthandProps(props, widgetShorthandProps),
+  );
+
+  return state;
+};
+```
+
+### Pros and Cons
+
+- **Pros**
+  - Less chance to make mistakes
+  - Better dev-time type checking
+- **Cons**
+  - Helper types can make the resulting `State` type harder to understand for a dev who's unfamiliar with what they're doing
+
+## Discarded Solutions
+
+N/A
+
+## Open Issues
+
+In desigining the `ComponentState` type, I opted for explicitly naming the ShorthandProps and DefaultedProps, in a nonconventional syntax:
+
+```typescript
+export type WidgetState = ComponentState<
+  WidgetProps,
+  { ShorthandProps: 'icon' | 'text' },
+  { DefaultedProps: 'text' | 'propWithDefaultValue' }
+>;
+```
+
+Is this syntax weird? The alternative is to have two normal type params, and rely on comments:

```

Comment:
I think this is smth that will need to update all usages at once and the scaffolding too, however after that I don't many ppl will really question it

Reply:
And on second thought, without an example it wouldn't be easy to infer the need of the nested objects, I think appropriate generic naming would be enough as far as compilation/intellisense is concerned

Reply:
Yeah, thinking about it more I agree that the extra objects aren't really necessary, especially if we add the comments to the boilerplate code.

Reply:
Updated to remove nested types